File: npcsh/version_check.py
import json
import os
import shutil
import subprocess
import threading
import time
from pathlib import Path
from urllib.request import urlopen, Request
from urllib.error import URLError, HTTPError

from npcsh.npcsh import VERSION

_CACHE_PATH = Path.home() / ".npcsh" / ".version_check"
_CACHE_TTL_SECONDS = 86400


class VersionCheck:
    """Check for newer versions of npcsh across pip, cargo, and brew."""

    def __init__(self):
        self.current = VERSION
        self.available = {}
        self.detected_source = self._detect_source()

    def _detect_source(self):
        """Figure out how npcsh was installed."""
        try:
            import npcsh
            pkg_dir = os.path.dirname(os.path.abspath(npcsh.__file__))
            if "site-packages" in pkg_dir or "dist-packages" in pkg_dir:
                return "pip"
        except Exception:
            pass

        binary = shutil.which("npcrs") or shutil.which("npcsh")

        brew_paths = [
            "/opt/homebrew",
            "/usr/local",
            "/home/linuxbrew",
        ]
        if binary:
            for bp in brew_paths:
                if binary.startswith(bp):
                    return "brew"
            return "cargo"

        return "pip"

    def _cached_result(self):
        """Return cached (source, latest, checked_at) if still fresh."""
        if not _CACHE_PATH.exists():
            return None
        try:
            data = json.loads(_CACHE_PATH.read_text())
            checked_at = data.get("checked_at", 0)
            if time.time() - checked_at < _CACHE_TTL_SECONDS:
                return data
        except (json.JSONDecodeError, OSError):
            pass
        return None

    def _save_cache(self, data):
        """Write check result to cache file."""
        try:
            _CACHE_PATH.parent.mkdir(parents=True, exist_ok=True)
            _CACHE_PATH.write_text(json.dumps(data))
        except OSError:
            pass

    def _pip_latest(self):
        """Fetch latest version from PyPI."""
        req = Request(
            "https://pypi.org/pypi/npcsh/json",
            headers={"User-Agent": "npcsh-version-check"},
        )
        try:
            with urlopen(req, timeout=5) as resp:
                data = json.loads(resp.read().decode())
                return data["info"]["version"]
        except (URLError, HTTPError, json.JSONDecodeError, KeyError):
            return None

    def _cargo_latest(self):
        """Fetch latest version from crates.io."""
        req = Request(
            "https://crates.io/api/v1/crates/npcsh",
            headers={"User-Agent": "npcsh-version-check"},
        )
        try:
            with urlopen(req, timeout=5) as resp:
                data = json.loads(resp.read().decode())
                return data["crate"]["max_stable_version"]
        except (URLError, HTTPError, json.JSONDecodeError, KeyError):
            return None

    def _brew_latest(self):
        """Fetch latest version from Homebrew formula."""
        try:
            result = subprocess.run(
                ["brew", "info", "npcsh"],
                capture_output=True,
                text=True,
                timeout=5,
            )
            if result.returncode == 0:
                first_line = result.stdout.splitlines()[0]
                if "stable" in first_line:
                    parts = first_line.split()
                    for i, p in enumerate(parts):
                        if p == "stable" and i + 1 < len(parts):
                            return parts[i + 1]
        except (FileNotFoundError, subprocess.TimeoutExpired):
            pass

        req = Request(
            "https://raw.githubusercontent.com/NPC-Worldwide/homebrew-npcsh/main/Formula/npcsh.rb",
            headers={"User-Agent": "npcsh-version-check"},
        )
        try:
            with urlopen(req, timeout=5) as resp:
                text = resp.read().decode()
                for line in text.splitlines():
                    if "version" in line and '"' in line:
                        start = line.find('"') + 1
                        end = line.find('"', start)
                        if end > start:
                            return line[start:end]
        except (URLError, HTTPError):
            pass

        return None

    def _parse_version(self, v):
        """Simple tuple version parser for comparison."""
        if not v:
            return (0, 0, 0)
        parts = v.lstrip("v").split(".")
        try:
            return tuple(int(p) for p in parts[:3])
        except ValueError:
            return (0, 0, 0)

    def check(self):
        """Run version checks and return update info if any."""
        cached = self._cached_result()
        if cached:
            self.available = cached.get("available", {})
        else:
            latest = {
                "pip": self._pip_latest(),
                "cargo": self._cargo_latest(),
                "brew": self._brew_latest(),
            }
            self.available = latest
            self._save_cache({
                "available": latest,
                "checked_at": time.time(),
            })

        source = self.detected_source
        latest = self.available.get(source)
        if not latest:
            return None

        if self._parse_version(latest) > self._parse_version(self.current):
            return {
                "source": source,
                "current": self.current,
                "latest": latest,
                "command": self._update_command(source),
            }
        return None

    def _update_command(self, source):
        if source == "pip":
            return "pip install --upgrade npcsh"
        if source == "cargo":
            return "cargo install npcsh --force"
        if source == "brew":
            return "brew upgrade npcsh"
        return ""

    def check_async(self, callback):
        """Run check in background and call callback with result."""
        def _run():
            result = self.check()
            if result:
                callback(result)

        thread = threading.Thread(target=_run, daemon=True)
        thread.start()

File: rust/src/version_check.rs
use std::path::PathBuf;
use std::time::{SystemTime, UNIX_EPOCH};

const CACHE_TTL_SECONDS: u64 = 86400;
const GITHUB_RELEASES_URL: &str = "https://api.github.com/repos/NPC-Worldwide/npcsh/releases/latest";
const CRATES_URL: &str = "https://crates.io/api/v1/crates/npcsh";
const BREW_FORMULA_URL: &str = "https://raw.githubusercontent.com/NPC-Worldwide/homebrew-npcsh/main/Formula/npcsh.rb";

#[derive(Debug, Clone)]
pub struct UpdateInfo {
    pub source: &'static str,
    pub current: String,
    pub latest: String,
    pub command: &'static str,
}

fn cache_path() -> PathBuf {
    std::path::PathBuf::from(shellexpand::tilde("~/.npcsh").as_ref()).join(".version_check.json")
}

fn now_secs() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs()
}

fn detect_source() -> &'static str {
    let exe = std::env::current_exe().ok();
    let exe_str = exe.as_ref().map(|p| p.display().to_string()).unwrap_or_default();

    let brew_roots = ["/opt/homebrew", "/usr/local", "/home/linuxbrew"];
    for root in &brew_roots {
        if exe_str.starts_with(root) {
            return "brew";
        }
    }

    if exe_str.contains(".cargo") || exe_str.contains("cargo-install") {
        return "cargo";
    }

    "binary"
}

fn read_cache() -> Option<(String, u64)> {
    let path = cache_path();
    let text = std::fs::read_to_string(&path).ok()?;
    let value: serde_json::Value = serde_json::from_str(&text).ok()?;
    let checked_at = value.get("checked_at")?.as_u64()?;
    if now_secs() - checked_at >= CACHE_TTL_SECONDS {
        return None;
    }
    let latest = value.get("latest")?.as_str()?.to_string();
    Some((latest, checked_at))
}

fn write_cache(latest: &str) {
    let path = cache_path();
    let _ = std::fs::create_dir_all(path.parent().unwrap_or(&path));
    let value = serde_json::json!({
        "latest": latest,
        "checked_at": now_secs(),
    });
    let _ = std::fs::write(path, value.to_string());
}

async fn fetch_github_latest(client: &reqwest::Client) -> Option<String> {
    let resp = client
        .get(GITHUB_RELEASES_URL)
        .header("User-Agent", "npcsh-version-check")
        .timeout(std::time::Duration::from_secs(5))
        .send()
        .await
        .ok()?;
    let json: serde_json::Value = resp.json().await.ok()?;
    json.get("tag_name")?.as_str().map(|t| t.trim_start_matches('v').to_string())
}

async fn fetch_crates_latest(client: &reqwest::Client) -> Option<String> {
    let resp = client
        .get(CRATES_URL)
        .header("User-Agent", "npcsh-version-check")
        .timeout(std::time::Duration::from_secs(5))
        .send()
        .await
        .ok()?;
    let json: serde_json::Value = resp.json().await.ok()?;
    json.get("crate")?
        .get("max_stable_version")?
        .as_str()
        .map(String::from)
}

async fn fetch_brew_latest(client: &reqwest::Client) -> Option<String> {
    let resp = client
        .get(BREW_FORMULA_URL)
        .header("User-Agent", "npcsh-version-check")
        .timeout(std::time::Duration::from_secs(5))
        .send()
        .await
        .ok()?;
    let text = resp.text().await.ok()?;
    for line in text.lines() {
        if line.trim().starts_with("version") {
            if let Some(start) = line.find('"') {
                if let Some(end) = line[start + 1..].find('"') {
                    return Some(line[start + 1..start + 1 + end].to_string());
                }
            }
        }
    }
    None
}

fn parse_version(v: &str) -> Option<semver::Version> {
    semver::Version::parse(v.trim_start_matches('v')).ok()
}

fn update_command(source: &str) -> &'static str {
    match source {
        "cargo" => "cargo install npcsh --force",
        "brew" => "brew upgrade npcsh",
        _ => "curl -fsSL https://enpisi.com/install-npcsh.sh | sh",
    }
}

pub async fn check_version(client: &reqwest::Client, current: &str) -> Option<UpdateInfo> {
    let source = detect_source();

    let latest = if let Some((cached, _)) = read_cache() {
        cached
    } else {
        let latest = match source {
            "cargo" => fetch_crates_latest(client).await,
            "brew" => fetch_brew_latest(client).await,
            _ => fetch_github_latest(client).await,
        }
        .unwrap_or_default();
        if !latest.is_empty() {
            write_cache(&latest);
        }
        latest
    };

    if latest.is_empty() {
        return None;
    }

    let current_ver = parse_version(current)?;
    let latest_ver = parse_version(&latest)?;

    if latest_ver > current_ver {
        Some(UpdateInfo {
            source,
            current: current.to_string(),
            latest,
            command: update_command(source),
        })
    } else {
        None
    }
}

pub fn format_update_notice(info: &UpdateInfo) -> String {
    format!(
        "  \x1b[33mUpdate available:\x1b[0m {} v{} → v{}\n  Run: \x1b[36m{}\x1b[0m",
        info.source, info.current, info.latest, info.command
    )
}
